Mathlib/Tactic/Setm.lean

```diff
@@ -0,0 +1,112 @@
+/-
+Copyright (c) 2023 Gareth Ma. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Gareth Ma
+-/
+import Lean
+import Mathlib.Tactic.DefEqTransformations
+
+/-!
+# The `setm` tactic
+
+The `setm` tactic ("`set` with matching") is used for introducing `let` declarations representing
+subexpressions of the goal or in the types of local hypotheses.
+
+For example, if the goal is `⊢ (x + 5) ^ 2 + (2 * y + x) * (x + 5) + 3 = 28` (with `x y : ℕ`), then
+`setm ?a ^ 2 + ?b * ?a + 3 = 28` would introduce `a : ℕ := x + 5` and `b : ℕ := 2 * y + 5` into the
+context and change the goal to `a ^ 2 + b * a + 3 = 28`
+-/
+
+namespace Mathlib.Tactic
+open Lean Parser Tactic Elab Tactic Meta
+
+initialize registerTraceClass `Tactic.setm
+
+/-- This is the core to the `setm` tactic. It takes an expression `e` and pattern `stx` containing
+named holes whose form should match `e`. For each named hole in `stx`, we create a `let` declaration
+of the same name whose value is filled by the corresponding subexpression of `e`. We then return
+the expression `e` with its subexpressions replaced by these named variables. -/
+def setMCore (e : Expr) (stx : TSyntax `term) : TacticM Expr := withMainContext do
+  let (origPattern, mvarIds) ← elabTermWithHoles stx none (←getMainTag) (allowNaturalHoles := true)
+  /- Named holes are by default syntheticOpaque and not assignable, so we change that -/
+  mvarIds.forM fun mvarId => mvarId.setKind .natural
+  trace[Tactic.setm] "origPattern : {← ppExpr origPattern}"
+
+  /- Create new placeholder mvars -/
+  let mvarIdsPairs ← mvarIds.mapM
+    (fun x => return (x, ←mkFreshExprMVar none (userName := (← x.getDecl).userName)))
+  let mvarIdsMap : HashMap MVarId Expr := .ofList mvarIdsPairs
+
+  /- newPattern is the placeholder pattern with a bunch of placeholder mvars -/
+  let newPattern := origPattern.replace fun x => match x with
+    | .mvar x => mvarIdsMap.find? x
+    | _ => none
+
+  /- We let isDefEq match the given pattern -/
+  let mctx ← getMCtx
+  let mdecls := mctx.decls
+  if ←isDefEq e newPattern then
+    /- We iterate over the (old, new) vars -/
+    mvarIdsPairs.forM fun (mvarIdOld, mvarExprNew) => do
+      let mvarIdNew := mvarExprNew.mvarId!
+      match (mdecls.find! mvarIdNew).userName with
+      | Name.anonymous => return ()
+      | name =>
+        let a := mkIdent name
+        let ha := mkIdent $ name.appendBefore "h"
```

Comment:
Stylistic convention for mathlib: always use `<|` for function application, never `$`! :)
```suggestion
       let ha := mkIdent <| name.appendBefore "h"
```